### ISSUE TITLE
False positive for type check on reusable workflow input?

### ISSUE THREAD

**Issue by bewuethr** [rhysd/actionlint#225]
I have a reusable workflow with inputs:

```yaml
name: Lint code base

on:
  workflow_call:
    inputs:
      validate-markdown:
        description: Run the Markdown linter
        type: string
        required: false
        default:
      validate-bash:
        description: Run ShellCheck
        type: string
        required: false
        default:

jobs:
  lint:
    name: Lint code base
    runs-on: ubuntu-20.04
    steps:
      - name: Check out repository
        uses: actions/checkout@v3.1.0

      - name: Lint code base
        uses: github/super-linter/slim@v4.9.7
        env:
          GITHUB_TOKEN: ${{ github.token }}
          VALIDATE_MARKDOWN: ${{ inputs.validate-markdown }}
          VALIDATE_BASH: ${{ inputs.validate-bash }}
```

and a workflow that calls that reusable workflow:

```yaml
name: Lint code base

on:
  workflow_dispatch:

jobs:
  lint:
    name: Lint code base
    uses: ./.github/workflows/linter.yml
    with:
      validate-markdown: true
```

The super-linter action doesn't allow setting some linters to `false` and some to `true`, so I use strings for the input type and use the empty string as a default. Super-linter then ignores the linters for which the env variable is empty, and handles `true` as a string just fine.

Now, actionlint complains about a type mismatch:

```console
$ actionlint linter-caller.yml 
linter-caller.yml:11:26: input "validate-markdown" is typed as string by reusable workflow "./.github/workflows/linter.yml". bool value cannot be assigned [expression]
   |
11 |       validate-markdown: true
   |                          ^~~~
```

Fair enough. I'll quote the parameter then:

```yaml
name: Lint code base

on:
  workflow_dispatch:

jobs:
  lint:
    name: Lint code base
    uses: ./.github/workflows/linter.yml
    with:
      validate-markdown: 'true'
```

but I still get the complaint:

```console
$ actionlint linter-caller.yml 
linter-caller.yml:11:26: input "validate-markdown" is typed as string by reusable workflow "./.github/workflows/linter.yml". bool value cannot be assigned [expression]
   |
11 |       validate-markdown: 'true'
   |                          ^~~~~~
```

Isn't that a false positive? The parameter should be considered a string when it's quoted, but it isn't? The workflow executes fine.

**Comment by bewuethr** [comment]
Using `${{ 'true' }}` works.

Is that correct behaviour from the parser? How would one force interpretation of `true` as a string generally in YAML, say, 1.2? The behaviour you describe (quotes are stripped) differs from the behaviour of, e.g., yq:

For this input

```yaml
a: true
b: 'true'
c: string
d: 'string'
e: "string"
```

`yq -P` (drop all unnecessary quotes) produces

```yaml
a: true
b: "true"
c: string
d: string
e: string
```

**Comment by pinkwah** [comment]
We encountered this today, with both a false positive and a false negative, where `actionlint` greatly diverges from behaviour expected from a compliant YAML parser.

In particular, using the examples above, we noticed that:
1. `validate-markdown: "true"` and `validate-markdown: !!str true` are string by YAML standards, but actionlint is incorrectly reporting that it's boolean. (This is the false positive case that is discussed here)
2. `validate-markdown: "True"` is a string by YAML and actionlint accepts it. (This works)
3. `validate-markdown: True` is a boolean by YAML, but actionlint accepts it because it incorrectly thinks it's a string. (This is a false negative)

If this is some weird behaviour that Github Actions' YAML parser has it should say so in the error message. If it's actionlint specifically then there is a bug in your YAML parser.

Here's an online example for how it's supposed to behave: https://jsonformatter.org/yaml-to-json/b610da

**Comment by LongBeachHXC** [comment]
Yeahhh, I'm just running into this today.

I'm getting around it by putting in an ignore flag.

This seems like an issue with actionlint eh? I'm clearly passing valid yaml strings and actionlint interpreting them as boolean, this smells like a bug.
